Interior penalty HDG

lindsayad opened the pull request:
Whereas https://github.com/idaholab/moose/pull/26477 is a mixed (value and gradient nonlinear variables) HDG method, this  is an interior penalty (no gradient nonlinear variable) one

Closes #30038


- [x] block restriction
- [x] displaced mesh
- [x] h-adaptivity following https://github.com/libMesh/libmesh/pull/4184. Added testing in lindsayad/moose@ffdc73be9f1 in #30357
- [x] explicit time integration: added error. Lagrange multipliers/face variables do not have time derivatives

lindsayad committed:
Implement interior penalty HDG method

Closes #30038

lindsayad committed:
Add ADHDGAdvection

lindsayad committed:
HDGDiffusionSideBC -> ADHDGSideDirichletBC

lindsayad committed:
Add simple 1d advection test

lindsayad committed:
Working advection-diffusion mms tests

lindsayad committed:
Add some more flexibility to ADHDGAdvection

lindsayad committed:
Add more flexibility to ADHDGAdvectionDirichletBC

This should probably be renamed at this point

lindsayad committed:
Add a 'IsMatrixSymmetric' postprocessor

lindsayad committed:
Successfully created symmetric ip-hdg discretization for stokes

lindsayad committed:
IP-HDG NS lid driven test

lindsayad committed:
lid-driven mms test for ip-hdg

lindsayad committed:
Respect user request to ignore matrix zero entries

lindsayad committed:
Add symmetric Stokes test

lindsayad committed:
Matrix testing

Currently the product and jump matrices don't match

lindsayad committed:
More flexibility in matrix printing problem class

- Allow not passing pressure_bar
- Allow not passing jump matrices

This is in preparation for some testing on Taylor-Hood

lindsayad committed:
IPHDGNavierStokesProblem -> PrintMatricesNSProblem

lindsayad committed:
Add capability for testing matrix equivalence

lindsayad committed:
Add *.mat to .gitignore

lindsayad committed:
Try to invert pressure mass matrix

GiudGiud commented:
did you think of how to differentiate them? Like prefix-wise maybe HDGIP ?

lindsayad committed:
View and load using PETSc interfaces

lindsayad commented:
I haven't. I'll give it some good thought at some point. Still heavily in the research phase

lindsayad committed:
With matching order p_bar, matrices are the same, woohoo

lindsayad committed:
Run test on larger mesh (10x10)

lindsayad committed:
Make IsMatrixSymmetric more robust

lindsayad committed:
Ensure that jump and vel-pb still equiv when h != 1

lindsayad committed:
Add eigenvalue computations

lindsayad committed:
Add NumZeroEigenvalues postprocessor

lindsayad committed:
Cleanup matrix test to focus on u-pb coupling

We don't care about the volumetric pressure dofs for our IP-HDG
scheme when constructing our grad-div perturbation analog

lindsayad committed:
Add gamma coefficient to our grad-div kernels

lindsayad committed:
Non-Reynolds robust IP-HDG field split input

lindsayad committed:
Some interesting matrix results

- With monomials for the interior velocity space, the grad-div FE form
  and the (up * M_p^-1 * pu) matrix are equivalent which is awesome.
  However, the corresponding div-grad form, e.g. (pu * M_u^-1 * up) is
  singular which is not awesome
- On the other hand, for Hierarchic/Lagrange, the grad-div FE form
  and the (up * M_p^-1 * pu) matrix are not equivalent which is not awesome.
  However, the corresponding div-grad form, e.g. (pu * M_u^-1 * up) is
  not singular which is awesome

How do we get the best of both worlds? Is there a bug somewhere?

lindsayad committed:
Add second order MMS testing

lindsayad committed:
Beef up matrix testing

The key change to get all I wanted:
- equiv grad-div FE form and B_up M_p^-1 B_pu
- equiv jump FE form and B_upb M_pb^-1 B_pbu
- nonsingular div-grad forms for each

was to use simplices instead of tensor products. I do not know why

lindsayad committed:
Illustrate very effective pressure mass matrix preconditioning with FSP

lindsayad committed:
Error if PrintMatricesNSProblem is used with comm size > 1

lindsayad committed:
Add combined grad_div and jump fe term matrix testing

This shows a single zero eigenvalue in the corresponding matrix:

B_pall_u * M_u^-1 * B_u_pall

which I'm guessing is due to the pressure null space

lindsayad committed:
Use auto after pointer cast

lindsayad committed:
Don't use singleton mutex to lock Jacobian additions

Otherwise error message reporting that could happen up the stack
will deadlock because it also tries to lock the singleton mutex

lindsayad committed:
Finally, a static condensation test

lindsayad committed:
Remove redundant protected in FEProblemBase

lindsayad committed:
Add protective error for nl_sys + l_sys

lindsayad committed:
Don't fill diagonal when using AD with static condensation

lindsayad committed:
Add static condensation MMS test

lindsayad committed:
PetscMatrix -> PetscMatrixBase

lindsayad committed:
Revamp IPHDG design to allow for modular term-by-term construction

Also adapt IPHDGKernel to HDGKernel

lindsayad committed:
Add statically condensed simple ip diffusion test

lindsayad committed:
Move all hdg tests under hdgkernels

lindsayad committed:
Create StaticCondensation preconditioner

lindsayad committed:
File rename to delineate LHDG and IPHDG as subclasses of HDG

lindsayad committed:
Adapt NS ip-hdg input to use HDGKernels

lindsayad committed:
Add static condensation test

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2668026/), step Clang format on d1db7b4 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27188/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27188/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format f5a14f390f39cec590003be860130561954c6b99`

lindsayad committed:
clang 16 formatting

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2729777/), step Docs: sync website on 0df7822 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/27188/site)

This comment will be updated on new commits.

lindsayad committed:
Remove remainder of AddHDGBCAction

lindsayad committed:
Restore phi_size member var in Kernel

lindsayad committed:
Remove now extraneous error about mixing linear and nl systems

lindsayad committed:
Add back mass and jump DG kernels to matrix testing input

lindsayad committed:
Adapt NS to HDG updates

lindsayad committed:
Remove LIBMESH_CHKERR

lindsayad committed:
Remove AreMatricesTheSame::initialize and use PetscSupport util

lindsayad committed:
Fundamentally change HDGKernels to be like reg kernels

lindsayad committed:
Add PetscSupport::createMatrixFromFile function

lindsayad committed:
Ensure that DGKernels and HDGKernels can co-execute

lindsayad committed:
More efficient way of coexisting DG and HDG

One less virtual call at least. Of course this commit title makes
me think about how we have inefficiencies around these two sets of
objects possibly calling reinit on the same face

lindsayad committed:
Fix missing ksp_view_pmat argument

lindsayad committed:
Restore strong FSP behavior for grad-div augmented IP-HDG

lindsayad committed:
Slightly bump rel_err for ldg channel test

lindsayad committed:
Update appsyntax python test for additional matrix_only parameter

lindsayad committed:
Add missing documentation

lindsayad committed:
Fix unity build failure

lindsayad committed:
Remove inflow condition assertion for MMS testing

lindsayad committed:
StaticCondensation doesn't support multiple threads

lindsayad committed:
Inline header methods

lindsayad committed:
Switch to TRI6 for tests

- QUAD9 appears to be unstable. I think this might be generally true
  for HDG
- Also, the condensed matrix is now symmetric with TRI6?
- Finally, we no longer use ParsedMaterial so we can remove the installation_type
  restriction
- Don't run FSP test with valgrind; it's too slow

moosebuild commented:
Job [Modules parallel](https://civet.inl.gov/job/2761326/) on 019946e : invalidated by @lindsayad

lindsayad committed:
Don't need ComputeJacobianThread::_add_mutex anymore

It was insufficient to prevent thread races, so now we have libmesh/libmesh#4113

lindsayad commented:
I'm used to waiting on this PR 😆 😢 See libmesh/libmesh#4113

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit b30db56c835158b7f36397e4339c493a28592480
+Subproject commit 1e10fa9fb72d10531a6128f821f8d11f8032882c
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2767595/), step Versioner verify on 741fb45 wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner versions

Found 10 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version            | to version         |
|-----------|----------|---------|-----------|--------------------|--------------------|
| libmesh   | CHANGE   | dabc922 | 83f53c8   | 2025.02.25 build 0 | 2025.03.22 build 0 |
| moose-dev | CHANGE   | fcaa73c | 2aebace   | 2025.03.03         | 2025.03.22         |

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2767863/), step Versioner verify on 47d8dde wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner versions

Found 10 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version            | to version         |
|-----------|----------|---------|-----------|--------------------|--------------------|
| libmesh   | CHANGE   | dabc922 | 83f53c8   | 2025.02.25 build 0 | 2025.03.22 build 0 |
| moose-dev | CHANGE   | fcaa73c | 2aebace   | 2025.03.03         | 2025.03.22         |

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2769839/), step Versioner verify on ac6d2cd wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner versions

Found 10 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version            | to version         |
|-----------|----------|---------|-----------|--------------------|--------------------|
| libmesh   | CHANGE   | c2aeeb2 | 83f53c8   | 2025.03.06 build 0 | 2025.03.22 build 0 |
| moose-dev | CHANGE   | ca0a369 | 2aebace   | 2025.03.06         | 2025.03.22         |

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2769881/), step Generate coverage on ac6d2cd wanted to post the following:


## Framework coverage

No coverage report(s) were found for the base commit 55c8eb2d3e.
[Full coverage report](https://mooseframework.inl.gov/docs/PRs/27188/coverage/framework)


## Modules coverage

No coverage report(s) were found for the base commit 55c8eb2d3e.

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/27188/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/27188/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/27188/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/27188/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/27188/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/27188/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/27188/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/27188/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/27188/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/27188/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/27188/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/27188/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/27188/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/27188/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/27188/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/27188/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/27188/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/27188/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/27188/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/27188/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/27188/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/27188/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/27188/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/27188/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/27188/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/27188/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/27188/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/27188/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/27188/coverage/xfem)

</details>



This comment will be updated on new commits.

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit c4791b4af32e3864d5160663d2c9e61989fe8d69
+Subproject commit 1e10fa9fb72d10531a6128f821f8d11f8032882c
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2773165/), step Versioner verify on 516a105 wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner versions

Found 10 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version            | to version         |
|-----------|----------|---------|-----------|--------------------|--------------------|
| libmesh   | CHANGE   | c2aeeb2 | 83f53c8   | 2025.03.06 build 0 | 2025.03.22 build 0 |
| moose-dev | CHANGE   | ca0a369 | 2aebace   | 2025.03.06         | 2025.03.22         |

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2773229/), step Versioner verify on 5fa1072 wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner versions

Found 10 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version            | to version         |
|-----------|----------|---------|-----------|--------------------|--------------------|
| libmesh   | CHANGE   | bc5c687 | 420783d   | 2025.03.06 build 1 | 2025.03.25 build 0 |
| moose-dev | CHANGE   | 00b281f | f48fbee   | 2025.03.24         | 2025.03.25         |

lindsayad committed:
Give AdvectionIPHDGAssemblyHelper capabilities of ADHDGAdvectionDirichletBC

lindsayad committed:
Update lid-driven-fsp input file to use AdvectionIPHDGAssemblyHelper

lindsayad committed:
Revert "Update lid-driven-fsp input file to use AdvectionIPHDGAssemblyHelper"

This reverts commit 05df87501c404c138984c0ed0bd1aaa5b213ea87.

lindsayad committed:
Rewrite IPHDG advection Dirichlet method to match ADHadvectionAdvectionDirichletBC

lindsayad committed:
Revert "Rewrite IPHDG advection Dirichlet method to match ADHadvectionAdvectionDirichletBC"

This reverts commit c6eab4d1d82662348a0c870c7e06e833ff48562a.

lindsayad committed:
Revert "Give AdvectionIPHDGAssemblyHelper capabilities of ADHDGAdvectionDirichletBC"

This reverts commit feea28a87a699cbed303beb19b49346f66e5f816.

lindsayad committed:
Remove extraneous boundary conditions for IP-HDG

lindsayad committed:
Eliminate need for ADHDGAdvectionDirichletBC

lindsayad committed:
Remove unused objects or move them to where they're used

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2777451/), step Versioner verify on 3c0218c wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner versions

Found 10 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version            | to version         |
|-----------|----------|---------|-----------|--------------------|--------------------|
| libmesh   | CHANGE   | bc5c687 | 420783d   | 2025.03.06 build 1 | 2025.03.25 build 0 |
| moose-dev | CHANGE   | 00b281f | f48fbee   | 2025.03.24         | 2025.03.25         |

lindsayad committed:
Add documentation

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2782131/), step Versioner verify on 74f2473 wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner versions

Found 10 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version            | to version         |
|-----------|----------|---------|-----------|--------------------|--------------------|
| libmesh   | CHANGE   | bc5c687 | 420783d   | 2025.03.06 build 1 | 2025.03.25 build 0 |
| moose-dev | CHANGE   | 00b281f | f48fbee   | 2025.03.24         | 2025.03.25         |

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2792028/), step Clang format on 0a34b7f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27188/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27188/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a7b98aa272700a4a06703efd4da8a041454da026`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2792028/), step Versioner verify on 0a34b7f wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner versions

Found 10 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version            | to version         |
|-----------|----------|---------|-----------|--------------------|--------------------|
| libmesh   | CHANGE   | bc5c687 | 420783d   | 2025.03.06 build 1 | 2025.03.25 build 0 |
| moose-dev | CHANGE   | 00b281f | f48fbee   | 2025.03.24         | 2025.03.25         |

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2792473/), step Versioner verify on b0ed508 wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner versions

Found 10 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version            | to version         |
|-----------|----------|---------|-----------|--------------------|--------------------|
| libmesh   | CHANGE   | bc5c687 | 420783d   | 2025.03.06 build 1 | 2025.03.25 build 0 |
| moose-dev | CHANGE   | 00b281f | f48fbee   | 2025.03.24         | 2025.03.25         |

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2794058/), step Versioner verify on 572f438 wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner versions

Found 10 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version            | to version         |
|-----------|----------|---------|-----------|--------------------|--------------------|
| libmesh   | CHANGE   | bc5c687 | 420783d   | 2025.03.06 build 1 | 2025.03.25 build 0 |
| moose-dev | CHANGE   | 00b281f | f48fbee   | 2025.03.24         | 2025.03.25         |

moosebuild commented:
Job [Conda (Rocky)](https://civet.inl.gov/job/2794093/) on 572f438 : invalidated by @lindsayad

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2794963/), step Versioner verify on bf62c2c wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner versions

Found 10 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version            | to version         |
|-----------|----------|---------|-----------|--------------------|--------------------|
| libmesh   | CHANGE   | bc5c687 | 420783d   | 2025.03.06 build 1 | 2025.03.25 build 0 |
| moose-dev | CHANGE   | 00b281f | f48fbee   | 2025.03.24         | 2025.03.25         |

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2796028/), step Versioner verify on ad1132d wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner versions

Found 10 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version            | to version         |
|-----------|----------|---------|-----------|--------------------|--------------------|
| libmesh   | CHANGE   | bc5c687 | 420783d   | 2025.03.06 build 1 | 2025.03.25 build 0 |
| moose-dev | CHANGE   | 00b281f | f48fbee   | 2025.03.24         | 2025.03.25         |

moosebuild commented:
Job [Apptainer build profiling](https://civet.inl.gov/job/2796106/) on 5e2c527 : invalidated by @lindsayad

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2797262/), step Versioner verify on a14a17a wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner versions

Found 10 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version            | to version         |
|-----------|----------|---------|-----------|--------------------|--------------------|
| libmesh   | CHANGE   | bc5c687 | 420783d   | 2025.03.06 build 1 | 2025.03.25 build 0 |
| moose-dev | CHANGE   | 00b281f | f48fbee   | 2025.03.24         | 2025.03.25         |

moosebuild commented:
Job [Modules](https://civet.inl.gov/job/2797288/) on a14a17a : invalidated by @lindsayad

moosebuild commented:
Job [Modules recover](https://civet.inl.gov/job/2797290/) on a14a17a : invalidated by @lindsayad

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2797304/), step Verify coverage on a14a17a wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `contact` coverage rate (required: 87.0%)
- Failed to generate `electromagnetics` coverage rate (required: 94.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 84.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 85.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 81.0%)
- Failed to generate `geochemistry` coverage rate (required: 96.0%)
- Failed to generate `heat_transfer` coverage rate (required: 87.0%)
- Failed to generate `level_set` coverage rate (required: 85.0%)
- Failed to generate `misc` coverage rate (required: 32.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- Failed to generate `peridynamics` coverage rate (required: 77.0%)
- Failed to generate `phase_field` coverage rate (required: 85.0%)
- Failed to generate `porous_flow` coverage rate (required: 95.0%)
- Failed to generate `ray_tracing` coverage rate (required: 93.0%)
- Failed to generate `rdg` coverage rate (required: 63.0%)
- Failed to generate `reactor` coverage rate (required: 90.0%)
- Failed to generate `scalar_transport` coverage rate (required: 81.0%)
- Failed to generate `solid_mechanics` coverage rate (required: 84.0%)
- Failed to generate `solid_properties` coverage rate (required: 83.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)
- Failed to generate `subchannel` coverage rate (required: 87.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 88.0%)
- Failed to generate `xfem` coverage rate (required: 80.0%)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2802002/), step Versioner verify on 9c5e63f wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner versions

Found 10 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version            | to version         |
|-----------|----------|---------|-----------|--------------------|--------------------|
| libmesh   | CHANGE   | bc5c687 | 420783d   | 2025.03.06 build 1 | 2025.03.25 build 0 |
| moose-dev | CHANGE   | 00b281f | f48fbee   | 2025.03.24         | 2025.03.25         |

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2806017/), step Versioner verify on d9b47de wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner versions

Found 10 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version            | to version         |
|-----------|----------|---------|-----------|--------------------|--------------------|
| libmesh   | CHANGE   | bc5c687 | 420783d   | 2025.03.06 build 1 | 2025.03.25 build 0 |
| moose-dev | CHANGE   | 00b281f | f48fbee   | 2025.03.24         | 2025.03.25         |

moosebuild commented:
Job [Conda (ARM Mac)](https://civet.inl.gov/job/2806051/) on d9b47de : invalidated by @lindsayad

moosebuild commented:
Job [Controlled apps](https://civet.inl.gov/job/2806046/) on d9b47de : invalidated by @lindsayad

GiudGiud commented:
How much of the previous HDG are you removing? Can we still do diffusion? NS?

GiudGiud commented:
I can review if you want

GiudGiud commented on framework/include/systems/NonlinearSystem.h:
```diff
@@ -24,17 +24,14 @@
  * It is a part of FEProblemBase ;-)
  */
-class NonlinearSystem : public NonlinearSystemBase,
-                        public libMesh::NonlinearImplicitSystem::ComputePreCheck
```
where did the preCheck go? or deleted? I might see it later in the review

lindsayad replied:
We don't need it any more. That's one of the very nice changes in this PR. The nonlinear system and its residual correspond to all the degrees of freedom, both interior and face degrees of freedom, instead of previously where I put the interior degrees of freedom in an auxiliary system. Then a user can choose whether they want to statically condense out the interior degrees of freedom in the *preconditioner* (leaving only the face degrees of freedom) or not.

This means a user is also free now to use any SNES algorithm they like as opposed to being limited to line-search without the line search (meaning previously you could only use `-snes_type newtonls -snes_linesearch_type basic`)

lindsayad replied:
This is the key class for this change: https://mooseframework.inl.gov/docs/doxygen/libmesh/classlibMesh_1_1StaticCondensation.html

lindsayad commented:
I didn't remove any of the previous HDG capability. I added an "L" to denote the previously implemented "local" HDG method and an "IP" to denote the newly added interior penalty HDG method

I welcome any reviewers!

GiudGiud commented on framework/include/hdgkernels/IPHDGAssemblyHelper.h:
```diff
@@ -0,0 +1,149 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
```
```suggestion
//* https://www.mooseframework.inl.gov
```

GiudGiud commented on framework/include/hdgkernels/IPHDGAssemblyHelper.h:
```diff
@@ -0,0 +1,149 @@
+class TransientInterface;
+
+/**
+ * Implements all the methods for assembling a hybridized interior penalty discontinuous Galerkin
+ * (IPDG-H), which is a type of HDG method, discretization of the diffusion equation. These routines
```
```suggestion
 * Base class that declares all the methods for assembling a hybridized interior penalty discontinuous Galerkin
 * (IPDG-H), which is a type of HDG method, discretization of an equation. These routines
```

GiudGiud replied:
if this is actuallt just diffusion let's rename the class

GiudGiud commented on framework/include/interfaces/TaggingInterface.h:
```diff
@@ -30,6 +30,13 @@ class Assembly;
 template <typename T>
 InputParameters validParams();
 
+struct ADResidualsPacket
```
docstring

Packet has communication connotations for me.

GiudGiud commented on framework/include/loops/NonlinearThread.h:
```diff
@@ -152,9 +153,22 @@ class NonlinearThread : public ThreadedElementLoop<ConstElemRange>
+private:
+  /// Whether DG kernels should be executed for a given internal side. This can be false if HDG
+  /// kernels are present because in that case we visit the internal side from both sides and we
+  /// don't want to double execute DG kernels
```
the logic seems to not match this comment?
>   return _should_execute_dg || _hdg_warehouse->hasActiveBlockObjects(_subdomain, _tid);

or do you mean "should execute dg" is false but that's ok for `shouldComputeInternalSide` because we also check the warehouse and will execute anyway?

lindsayad replied:
I've tried to clarify the comment

GiudGiud reviewed (commented):
only looked at framework doco and headers. I ll probably do 2 or 3 more rounds to cover everything

lindsayad commented on framework/include/bcs/AdvectionIPHDGOutflowBC.h:
```diff
@@ -0,0 +1,34 @@
+   */
+  virtual void compute() override;
+
+  virtual AdvectionIPHDGAssemblyHelper & iphdgHelper() override { return *_iphdg_helper; }
```
https://en.wikipedia.org/wiki/Covariant_return_type

GiudGiud replied:
oh nice. I need to start doing this

lindsayad commented on framework/include/hdgkernels/DiffusionLHDGKernel.h:
```diff
@@ -0,0 +1,49 @@
+  const MooseArray<Real> & _JxW_face;
+
+  /// face normals
+  const MooseArray<Point> & _normals;
```
I would need a new class of residual objects then that are stored in warehouses and leveraged in `ComputeResidual/JacobianThread`. Because in order for these systems to be statically condensed we must be assembling on a per-element basis. DGKernels muck that up because they add into both element and neighbor residuals/Jacobians so I cannot use them

GiudGiud commented on modules/navier_stokes/include/postprocessors/AreMatricesTheSame.h:
```diff
@@ -0,0 +1,39 @@
+class PetscMatrix;
+}
+
+class AreMatricesTheSame : public GeneralPostprocessor
```
```suggestion
/**
 * Checks if two matrices are the same by comparing their coefficients
 */
class AreMatricesTheSame : public GeneralPostprocessor
```

GiudGiud commented on modules/navier_stokes/include/kernels/GradDiv.h:
```diff
@@ -0,0 +1,31 @@
+
+#include "ADKernel.h"
+
+class GradDiv : public ADKernel
```
docstring

GiudGiud commented on python/MooseDocs/test/extensions/test_appsyntax.py:
```diff
@@ -156,7 +156,7 @@ def testLatex(self):
         # Diffusion (Required + Optional + Advanced + Tagging) + 1
         # (corresponding to 'type')
-        self.assertSize(res, 18)
+        self.assertSize(res, 19)
```
doco not built rn. what did you add to diffusion?

lindsayad replied:
`matrix_only` was made non-private

GiudGiud commented on modules/navier_stokes/test/tests/finite_element/ins/hdg/ip/lid-driven/tests:
```diff
@@ -0,0 +1,94 @@
+    []
+  []
+  [condensed_symmetric]
+    requirement = "The condensed matrix shall be symmetric, although I don't understand why. Where did the advection go?"
```
this wont look good in the SQA doco

advection hits the non-condensed dofs right, and should not be symmetric

maybe tolerances?

GiudGiud commented on modules/navier_stokes/test/tests/finite_element/ins/hdg/ip/lid-driven/sc-lid-driven.i:
```diff
@@ -0,0 +1,216 @@
+[]
+
+[HDGKernels]
+  [momentum_x_convection]
```
```suggestion
  [momentum_x_advection]
```
convection is only heat/ body force driven

GiudGiud commented on modules/navier_stokes/test/tests/finite_element/ins/hdg/ip/lid-driven/gold/lid-driven-fsp_out.e:
let's hide the side hierarchic vsariables, not output right

GiudGiud commented on modules/navier_stokes/test/tests/finite_element/ins/hdg/ip/lid-driven/gold/lid-driven_out.e:
pressure has a 3e5 cell to cell jump in this file

lindsayad replied:
Really? How did you determine that? The pressure field looks pretty reasonable to me
<img width="694" alt="Screenshot 2025-05-21 at 2 03 08 PM" src="https://github.com/user-attachments/assets/881d55cb-1a52-4dbc-bc64-03adb6768844" />

grmnptr replied:
With constant lid velocity we have those, analytic solution is not known. We will add some hdg cases to the flow test bed, there we can compare with validation results.

GiudGiud commented on modules/navier_stokes/include/hdgkernels/NavierStokesStressIPHDGAssemblyHelper.h:
```diff
@@ -0,0 +1,69 @@
+
+/**
+ * Implements all the methods for assembling a hybridized interior penalty discontinuous Galerkin
+ * (IPDG-H), which is a type of HDG method, discretization of the diffusion equation. These routines
```
```suggestion
 * (IPDG-H), which is a type of HDG method, discretization of the NS equation. These routines
```

GiudGiud reviewed (changes requested):
Still have to go over the NS source still

lindsayad committed:
Add exact component to MassFluxPenaltyBC

lindsayad committed:
Don't warn users about setting certain PETSc options

These are false warnings for PETSc power users and it's not even
true that we ignore the passed options

lindsayad committed:
Documentation improvements

lindsayad committed:
Address first round of Guillaume review

Principally documentation improvements

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2881774/), step Clang format on 8de3101 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27188/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27188/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 083e818fdf10367a54e8ba2444e70a9b273db34b`

lindsayad committed:
Rename var parameters to be more consistent with IP-HDG and the rest of MOOSE

lindsayad committed:
_my -> _cached

lindsayad committed:
Guillaume review round 2

- Remove unnecessary includes
- Avoid unnecessary qp calculations in shape function for loops
- More use of required parameters to force user to think about what they're doing and to commit less errors

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2882350/), step Clang format on d078f2e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27188/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27188/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 083e818fdf10367a54e8ba2444e70a9b273db34b`

lindsayad committed:
Don't need to add moose variable dependency for _u_var/'variable'

lindsayad committed:
Reorder L-HDG diffusionloops to reduce FLOPS

lindsayad committed:
Rename newly renamed gradient_variable and face_variable in NS

lindsayad commented on framework/src/hdgkernels/DiffusionLHDGAssemblyHelper.C:
```diff
@@ -0,0 +1,407 @@
+                                                  DenseVector<Number> & vector_re)
+{
+  for (const auto i : index_range(vector_re))
+    for (const auto qp : make_range(qrule.n_points()))
```
I reordered the loops now to be consistent with IP-HDG

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2883214/), step Clang format on b2af9cb wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/27188/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/27188/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 083e818fdf10367a54e8ba2444e70a9b273db34b`

lindsayad commented on framework/src/loops/NonlinearThread.C:
```diff
@@ -252,6 +256,15 @@ NonlinearThread::onInternalSide(const Elem * elem, unsigned int side)
   }
+  if (_hdg_warehouse->hasActiveBlockObjects(_subdomain, _tid))
+  {
+    // Set up Sentinel class so that, even if reinitMaterialsFace() throws, we
```
It's in `prepareFace`. I will update the comment to be more helpful

lindsayad committed:
More helpful comment around sentinels in NonlinearThread

lindsayad commented on framework/src/postprocessors/IsMatrixSymmetric.C:
```diff
@@ -45,30 +52,45 @@ void
 
-  for (const auto i : make_range(sys_mat.row_start(), sys_mat.row_stop()))
-    for (const auto j : make_range(sys_mat.col_start(), sys_mat.col_stop()))
+  mat->get_transpose(*_mat_transpose);
```
Great question. I'll give it a try

lindsayad committed:
No need for _mat_transpose in IsMatrixSymmetric

lindsayad committed:
Address more review

- Remove 'MOOSE includes' comment
- cubit -> cubic
- Remove double parentheses in constructor

lindsayad commented on framework/src/preconditioners/MooseStaticCondensationPreconditioner.C:
```diff
@@ -0,0 +1,71 @@
+        _fe_problem, root_old_prefix + petsc_system + "_", root_new_prefix + petsc_system + "_");
+  };
+  change_prefix("ksp");
+  change_prefix("pc");
```
Good point. I've revamped this to blanketly apply the prefix to all options provided in this preconditioner's block. Importantly it also no longer applies to PETSc options provided elsewhere, like in the Executioner, which makes it so that you can set things like `-ksp_type preonly` for the global linear solve in the input file as opposed to needing to set it on the command line, which is something I often forget to do. And if I as the author forget to do it, then certainly many others would also

lindsayad committed:
Uniformly apply prefix in MooseStaticCondensationPreconditioner

lindsayad committed:
Ensure SC and non-SC solver options are being used

lindsayad committed:
Require users to specify functors in velocity HDG Dirichlet BCs

lindsayad committed:
NS documentation improvements

lindsayad commented on modules/navier_stokes/include/bcs/AdvectionIPHDGPrescribedFluxBC.h:
```diff
@@ -0,0 +1,29 @@
+#include "IPHDGPrescribedFluxBC.h"
+
+/**
+ * Implements an outflow boundary condition for use with a hybridized discretization of
```
Thus far I've used it exclusively for prescribing zero advective flux conditions at wall-type boundaries but you're right that it could just as well be used for weakly prescribing an inlet advection flux. Really an outlet boundary is the only place this *shouldn't* be used! I've updated the doc

lindsayad committed:
Rename matrix classes in a uniform way

lindsayad commented on modules/navier_stokes/include/bcs/NavierStokesStressIPHDGDirichletBC.h:
```diff
@@ -0,0 +1,30 @@
+ * Weakly imposes Dirichlet boundary conditions for a hybridized discretization of a Navier-Stokes
+ * equation stress term"
+ */
+class NavierStokesStressIPHDGDirichletBC : public IPHDGDirichletBC
```
I have added some now in `HDGKernels/index.md`

lindsayad committed:
Hide side_u output

lindsayad committed:
convection -> advection

lindsayad committed:
Fix symmetry check for IP-HDG NS

lindsayad committed:
Error in 3D in MassFluxPenaltyBC

lindsayad commented on modules/navier_stokes/src/bcs/MassFluxPenaltyBC.C:
```diff
@@ -0,0 +1,50 @@
+ADReal
+MassFluxPenaltyBC::computeQpResidual()
+{
+  const ADRealVectorValue soln_jump(_vel_x[_qp], _vel_y[_qp], 0);
```
I've added a Dirichlet value parameter so that this becomes the correct jump term

lindsayad commented on modules/navier_stokes/src/bcs/NavierStokesLHDGOutflowBC.C:
```diff
@@ -0,0 +1,210 @@
+  addJacobian(
+      _assembly, _p_lm_u_jac, _p_dof_indices, _lm_u_dof_indices, _pressure_var.scalingFactor());
+  addJacobian(
+      _assembly, _p_lm_v_jac, _p_dof_indices, _lm_v_dof_indices, _pressure_var.scalingFactor());
```
what do you mean? Like `-snes_test_jacobian`?

GiudGiud replied:
ah yeah that works for BCs. 
The one that does not is the analyze-jacobian.py right?

lindsayad commented on modules/navier_stokes/test/src/problems/PrintMatricesNSProblem.C:
```diff
@@ -0,0 +1,218 @@
+    {
+      Mat B, M, Minv;
+
+      // Create B
```
yes, I'll make that more clear. Very poor choice of variable name given that B typically represents the v-p coupling matrix

lindsayad committed:
Miscellaneous improvements

- Better documentation
- Error for objects that currently only work in 2D

lindsayad committed:
Clang formatting

lindsayad committed:
Set Preconditioning solve_type at construction time

lindsayad committed:
mooseframework.org -> mooseframework.inl.gov

GiudGiud commented on framework/include/bcs/IPHDGBC.h:
```diff
@@ -0,0 +1,55 @@
+  const Elem * _my_elem;
+
+  /// A cache variable to prevent multiple computations of Jacobians
+  unsigned int _my_side;
```
```suggestion
  unsigned int _cached_side;
```

GiudGiud reviewed (commented):
looked at most of the headers

GiudGiud reviewed (commented):
Done with framework/

so in light of the discussion on features, what would you consider to be reasonable things to check that they work with HDG? block restriction? displaced mesh? h-adaptivity? explicit time integration?

lindsayad commented on framework/src/hdgkernels/DiffusionIPHDGAssemblyHelper.C:
```diff
@@ -0,0 +1,111 @@
+
+#include "DiffusionIPHDGAssemblyHelper.h"
+#include "MooseTypes.h"
+#include "MooseVariableDependencyInterface.h"
+#include "MooseVariableScalar.h"
+#include "SystemBase.h"
+#include "MooseMesh.h"
+#include "MooseObject.h"
+#include "MaterialPropertyInterface.h"
```
I don't think I can remove all of these in all the helpers

GiudGiud replied:
no they dont have the same list but they do seem longer than needed

lindsayad committed:
more _my -> _cached

lindsayad committed:
Unnecessary include removal

lindsayad commented on framework/src/bcs/DiffusionLHDGDirichletBC.C:
```diff
@@ -0,0 +1,109 @@
+}
+
+void
+DiffusionLHDGDirichletBC::computeOffDiagJacobian(const unsigned int)
```
`computeOffDiagJacobian` is always the residual object entry point from the Jacobian `NonlinearThread` unless the coupling type is `COUPLING_DIAG`, corresponding to no preconditioning block. And since I error if full coupling is not requested, there is no chance of having the entry point be directly `computeJacobian`. Related code is
https://github.com/idaholab/moose/blob/7cc8bd7530c04b73ef16019251092c5a42403fdc/framework/src/systems/NonlinearSystemBase.C#L2877-L2907

GiudGiud replied:
ok that works
if doxygen lets you maybe re-do the docstring then?  They are from the RO base class and they don't match what these new classes do. 

computeJacobian is no longer
```
  /// Compute this object's contribution to the diagonal Jacobian entries
```
and computeOffDiagJacobian is not
```
   * Computes this object's contribution to off-diagonal blocks of the system Jacobian matrix
```
they both just do the full Jacobian.

lindsayad committed:
More and better English

- More precise error about specifying coupling on a per-system basis
- No reason for capitalized advection in doc string

lindsayad committed:
Remove duplicate _u_var from additionalVariablesCovered

lindsayad committed:
Add transient HDG test

lindsayad committed:
Better doxygen for element Jacobian methods for HDGKernel derived classes

lindsayad commented:
> so in light of the discussion on features, what would you consider to be reasonable things to check that they work with HDG? block restriction? displaced mesh? h-adaptivity? explicit time integration?

I really like the new explicit time integrator error. It catches bad uses right away
```
*** ERROR ***
CLI_ARGS:1.1:
The following error occurred in the TimeIntegrator 'ActuallyExplicitEuler:nl0' of type ActuallyExplicitEuler.

No objects contributing to the mass matrix were found for variable 'side_u'. Did you, e.g., forget a time derivative term?
```
I'll add a more HDG specific error

lindsayad committed:
Error if HDGKernel used with explicit time integration

GiudGiud commented:
What if you did have a time derivative?

lindsayad commented:
You should not have a time derivative. While the facet variables do represent the solution on faces, their ultimate jobs are as Lagrange multipliers enforcing flux constraints on the faces

GiudGiud commented:
ah ok I was thinking about a time derivative in general. But I understand this will be caught by the lack of derivative on the face term. Having an error works for me

GiudGiud commented on modules/navier_stokes/src/hdgkernels/NavierStokesLHDGAssemblyHelper.C:
```diff
@@ -0,0 +1,578 @@
+    if (_enclosure_lm_var)
+    {
+      for (const auto j : make_range(global_lm_p_jac.n()))
+        global_lm_p_jac(0, j) -= JxW[qp] * _scalar_phi[j][qp];
```
so this routine is assuming that p and v use the same shape functions right?

GiudGiud replied:
otherwise dp/dp_i would be a different phi

do we check for this somewhere?

GiudGiud commented on modules/navier_stokes/src/hdgkernels/NavierStokesLHDGAssemblyHelper.C:
```diff
@@ -0,0 +1,578 @@
+                                                     DenseMatrix<Number> & p_lm_u_vel_jac,
+                                                     DenseMatrix<Number> & p_lm_v_vel_jac)
+{
+  for (const auto i : make_range(p_lm_u_vel_jac.m()))
```
I d assert p_lm_u_vel_jac and p_lm_v_vel_jac same size

GiudGiud replied:
or not. I was just getting confused at first reading the loop ranges

GiudGiud replied:
it wont hurt though

GiudGiud commented on modules/navier_stokes/src/hdgkernels/NavierStokesStressIPHDGAssemblyHelper.C:
```diff
@@ -0,0 +1,92 @@
+
+  for (const auto i : index_range(_scalar_re))
+    for (const auto qp : make_range(_ip_qrule_face->n_points()))
+      _scalar_re(i) += _ip_JxW_face[qp] * _pressure_face_sol[qp] * _ip_normals[qp](_component) *
```
so if a user specified a dirichletBC on pressure, does that set `_pressure_face_sol` in the backend?

this is curiosity. A user should not set a dirichlet BC on pressure outside on a stress bdy/wall

lindsayad replied:
Yes. This would still have dependence on the pressure dofs but the pressure dof Jacobian rows would reflect the Dirichlet boundary condition

GiudGiud commented on modules/navier_stokes/src/dgkernels/MassFluxPenalty.C:
```diff
@@ -0,0 +1,71 @@
+  params.addRequiredParam<unsigned short>("component",
+                                          "The velocity component this object is being applied to");
+  params.addParam<Real>("gamma", 1, "The penalty to multiply the jump");
+  return params;
```
missing class description. 
maybe we should delete ADDGKernel's so we get warned of these things

GiudGiud reviewed (commented):
done with NS

lindsayad committed:
Fix block-restricted behavior for HDGKernels and add test

lindsayad committed:
Add displaced mesh test

lindsayad committed:
additionalVariablesCovered -> additionalROVariables

lindsayad committed:
Address review comments

- present -> presented
- Use ADFunctorInterface
- Error if a user attempts to use a non-Cartesian coordinate system

lindsayad committed:
Address review comments

- Error check for same finite element type for scalar field, vector
  field, and face variable finite element types
- Add missing class description

lindsayad committed:
Restrict SC pre tests to max_threads 1

lindsayad commented:
This should be ready for further review

GiudGiud reviewed (approved):
Resolved all my comments, all addressed. Looks good to me!

lindsayad commented:
Thanks for a truly great review. The question about features helped highlight several areas for improvement and bug fixes

GiudGiud commented:
patch is up
https://github.com/idaholab/moose/pull/30685

TMAP8 patch is easy if we dont like my moose change
